GeneralisedFilters/src/containers.jl

```diff
@@ -8,39 +8,69 @@
 A container for particle filters which composes the weighted sample into a distibution-like
 object, with the states (or particles) distributed accoring to their log-weights.
 """
-mutable struct ParticleDistribution{PT,WT<:Real}
+abstract type ParticleDistribution{PT} end
+
+Base.collect(state::PT) where {PT<:ParticleDistribution} = state.particles
+Base.length(state::PT) where {PT<:ParticleDistribution} = length(state.particles)
+Base.keys(state::PT) where {PT<:ParticleDistribution} = LinearIndices(state.particles)
+
+Base.iterate(state::ParticleDistribution, i) = Base.iterate(state.particles, i)
+Base.iterate(state::ParticleDistribution) = Base.iterate(state.particles)
+
+# not sure if this is kosher, since it doesn't follow the convention of Base.getindex
+Base.@propagate_inbounds Base.getindex(state::ParticleDistribution, i) = state.particles[i]
+
+mutable struct Particles{PT} <: ParticleDistribution{PT}
+    particles::Vector{PT}
+    ancestors::Vector{Int}
+end
+
+mutable struct WeightedParticles{PT,WT<:Real} <: ParticleDistribution{PT}
     particles::Vector{PT}
     ancestors::Vector{Int}
     log_weights::Vector{WT}
 end
-function ParticleDistribution(particles::Vector{PT}, log_weights::Vector{WT}) where {PT,WT}
+
+function Particles(particles::AbstractVector)
     N = length(particles)
-    return ParticleDistribution(particles, Vector{Int}(1:N), log_weights)
+    return Particles(particles, Vector{Int}(1:N))
 end
 
-StatsBase.weights(state::ParticleDistribution) = softmax(state.log_weights)
+function WeightedParticles(particles, log_weights)
+    N = length(particles)
+    return WeightedParticles(particles, Vector{Int}(1:N), log_weights)
+end
 
-Base.collect(state::ParticleDistribution) = state.particles
-Base.length(state::ParticleDistribution) = length(state.particles)
-Base.keys(state::ParticleDistribution) = LinearIndices(state.particles)
+StatsBase.weights(state::Particles) = fill(1 / length(state), length(state))
+StatsBase.weights(state::WeightedParticles) = softmax(state.log_weights)
 
-# not sure if this is kosher, since it doesn't follow the convention of Base.getindex
-Base.@propagate_inbounds Base.getindex(state::ParticleDistribution, i) = state.particles[i]
-# Base.@propagate_inbounds Base.getindex(state::ParticleDistribution, i::Vector{Int}) = state.particles[i]
+function update_weights(state::Particles, log_weights)
+    return WeightedParticles(state.particles, state.ancestors, log_weights)
+end
 
-function reset_weights!(state::ParticleDistribution{T,WT}) where {T,WT<:Real}
-    fill!(state.log_weights, zero(WT))
-    return state.log_weights
+function update_weights(state::WeightedParticles, log_weights)
```

Comment:
```suggestion
function update_weights!(state::WeightedParticles, log_weights)
```

By convention, functions that mutate their inputs should have their name end in `!`, per https://docs.julialang.org/en/v1/manual/style-guide/#bang-convention.

There may be more instances of this in the PR that I haven't noticed, just spotted this one.

Reply:
@mhauru I apologize for the untimely response, but there is actually good reason for this. Calls to `update_weights` given `::Particles` allocate a new `::WeightedParticles`, since we only learn of the type upon the first call to `SSMProblems.logdensity` on the observation process.

Ideally I would return a `::WeightedParticles` from the initial draw, but if the weight type varies we make way for type unstable code. There's no perfect solution here, but it's the only way to ensure type stability (to my knowledge).

Reply:
If some methods of the function mutate and others don't (this is what I understood from your comment) then you could call it `update_weights!!` according to the [BangBang](https://github.com/juliafolds/bangbang.jl) convention.

Reply:
I should have thought more seriously about `BangBang.jl`, this seems to be a very relevant convention for this PR.

Reply:
Should the `predict/update` interface also use `!!` since they mutate the particle container ?